crates/binjs_io/src/json/value.rs

```diff
@@ -0,0 +1,216 @@
+/// JS value in JSON.
+use TokenReaderError;
+
+use binjs_shared::SharedString;
+
+use std::borrow::Borrow;
+use std::collections::HashMap;
+use std::rc::Rc;
+
+/// JS value in JSON.
+pub enum Value {
+    String(SharedString),
+    Number(f64),
+    Object(Rc<HashMap<SharedString, Rc<Value>>>),
+    Array(Rc<Vec<Rc<Value>>>),
+    Bool(bool),
+    Null,
+}
+impl Value {
+    /// Get the string content from the value.
+    /// Fails if the value is not string.
+    pub fn as_string(&self) -> Result<SharedString, TokenReaderError> {
+        match self {
+            Value::String(s) => Ok(s.clone()),
+            _ => Err(TokenReaderError::GenericError(
+                "Expected a string".to_string(),
+            )),
+        }
+    }
+
+    /// Get the object content from the value.
+    /// Fails if the value is not object.
+    pub fn as_object(&self) -> Result<Rc<HashMap<SharedString, Rc<Value>>>, TokenReaderError> {
+        match self {
+            Value::Object(obj) => Ok(obj.clone()),
+            _ => Err(TokenReaderError::GenericError(
+                "Expected an object".to_string(),
+            )),
+        }
+    }
+}
+
+/// Methods to treat HashMap as JS Object.
+pub trait ObjectValue {
+    /// Get the property from the object.
+    /// Fails if the property doesn't exist.
+    fn get_property(&self, key: &'static str) -> Result<Rc<Value>, TokenReaderError>;
+
+    /// Get the property from the object and return bool or null content.
+    /// Returns Some(bool) if the property is bool.
+    /// Returns None if the property is null
+    ///
+    /// Fails if the property doesn't exist, or the property value is neither
+    /// bool nor null.
+    fn get_bool_or_null_property(
```

Comment:
Maybe `get_maybe_bool_property`, for consistency with other code (here and other methods below)?